### PR TITLE
ENH: Add placeholder stubs for all sub-modules

### PR DESCRIPTION
**Pull request by BvB93**
This pull request adds placeholder stub files for all (public) modules.

"Fixes" the issue encountered in https://github.com/numpy/numpy/issues/17103 by adding placeholder stubs for all currently untyped modules. 
Note that all objects within the respective modules are currently typed as `Any`, so this is more of a duct tape solution until proper stubs are added.

Examples
---------
How the submodules were identified:

``` python
In [1]: from types import ModuleType
   ...: from typing import Set
   ...: import numpy as np
   ...:
   ...: def get_submodules(super_module: ModuleType) -> Set[str]:
   ...:     """Return a set of all public sub-modules located in the passed `super_module`."""
   ...:     items = ((name, getattr(super_module, name)) for name in dir(super_module))
   ...:     ret = {name for name, obj in items if not name.startswith("_") and isinstance(obj, ModuleType)}
   ...:     return ret - {"sys", "os", "warnings", "math"}  # builtin module

In [2]: get_submodules(np)
Out[2]:
{'char',
 'compat',
 'core',
 'ctypeslib',
 'emath',
 'fft',
 'lib',
 'linalg',
 'ma',
 'matrixlib',
 'polynomial',
 'random',
 'rec',
 'testing',
 'version'}
```

### CODE REVIEW

File: numpy/warnings.pyi [hunk 1]
```diff
@@ -0,0 +1,4 @@
+from typing import Any
```
Comment (person142) [inline numpy/warnings.pyi:1]: This one is just the built in `warnings` module IIRC, so we don't want to include it?

File: numpy/tests/typing/reveal/modules.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+reveal_type(np.os)  # E: ModuleType
+reveal_type(np.polynomial)  # E: ModuleType
+reveal_type(np.random)  # E: ModuleType
+reveal_type(np.rec)  # E: ModuleType
+reveal_type(np.testing)  # E: ModuleType
+reveal_type(np.version)  # E: ModuleType
+reveal_type(np.warnings)  # E: ModuleType
+
+# TODO: Remove when annotations have been added to `np.linalg.norm`
+reveal_type(np.linalg.norm)  # E: Any
```
Comment (person142) [inline numpy/tests/typing/reveal/modules.py:25]: We should be able to deal with this one right now?

**Review by person142 — COMMENTED** [review]
I like this; in the future I think we could take it even further by having every attribute in the module enumerated and typed as `Any` (until we have stubs for it) and removing the `__getattr__`s. In practice I've noticed that partial stubs tend to cause accidental typos, i.e. things like

```python
def f(x: np.array): ...
```

It would be nice to have mypy catch those.

File: numpy/tests/typing/reveal/modules.py (hunk 2 above)
Comment (BvB93) [reply numpy/tests/typing/reveal/modules.py:25]: Indeed, this just serves as a test case to check that the module-level `__getattr__()` indeed returns `Any`.

File: numpy/warnings.pyi (hunk 1 above)
Comment (BvB93) [reply numpy/warnings.pyi]: You're absolutely right, fixed in https://github.com/numpy/numpy/pull/17104/commits/e91e31eab3f6a4af1498635c6c80d4ef4263e1d9.

File: numpy/__init__.pyi [hunk 3]
```diff
@@ -44,6 +44,28 @@ if sys.version_info >= (3, 8):
...
+    ma,
+    math,
+    matrixlib,
+    os,
+    polynomial,
+    random,
+    rec,
+    testing,
+    version,
+    warnings,
```
Comment (eric-wieser) [inline numpy/__init__.pyi:66]: You should be ignoring all the ones that are not part of the numpy installation
```suggestion
    char,
    compat,
    core,
    ctypeslib,
    emath,
    fft,
    lib,
    linalg,
    ma,
    matrixlib,
    polynomial,
    random,
    rec,
    testing,
    version,
```

Comment (BvB93) [reply numpy/__init__.pyi]: Removed the last ones as of https://github.com/numpy/numpy/pull/17104/commits/160a782381d2f723b0f1d836503e13eedeaaf7aa.

**Comment by bashtage** [comment]
Can these be smoothly extended, for example, could a part of random be typed without completing the full module and have that part correctly picked up?

**Comment by person142** [comment]
> Can these be smoothly extended, for example, could a part of random be typed without completing the full module and have that part correctly picked up?

Yup, in fact that's the current state of the main `numpy` namespace.

**Comment by BvB93** [comment]
> I like this; in the future I think we could take it even further by having every attribute in the module enumerated and typed as `Any` (until we have stubs for it) and removing the `__getattr__`s.

Good news: turns out the future is today! (see https://github.com/numpy/numpy/pull/17104/commits/4af8fd39bfa4a30e9d26e8021b42ffadf1f99398)

Explicit type annotations (_i.e._ `Any`) have now been added for all modules which define  `__all__`.
Note that a couple of them don't seem to define `__all__` (which they probably should); those were left untouched.

Examples
----------
The code used for grabbing all `__all__`s:

``` python
from types import ModuleType
from typing import Dict, Optional, List
import numpy as np

def get_submodule_all(super_module: ModuleType) -> Dict[str, Optional[List[str]]]:
    """Return a dict with all (public) sub-modules as keys and their respective `__all__` as values.

    A value is set to `None` if sub-module does not define `__all__`.
    """
    items = ((name, getattr(super_module, name)) for name in dir(super_module))
    superset = {name for name, obj in items if not name.startswith("_") and isinstance(obj, ModuleType)}
    sub_modules = superset - {"sys", "os", "warnings", "math"}  # builtin modules
    return {name: getattr(getattr(super_module, name), '__all__', None) for name in mod_set}
```

**Review by rgommers — COMMENTED** [review]
+1 good idea to add placeholder annotations for all public modules.

I would prefer if we didn't blindly annotate everything that doesn't start with an underscore (not sure if that's what you aimed for). Not typing things that are meant to be private, or things that are deprecated, would be better as a signal to users they're in the wrong place. 

See `numpy/tests/test_public_api.py` for things that look public but really aren't.

**Comment by bashtage** [comment]
I'm happy to help out but this has a new learning curve. 

I made an attempt on a single function from `random` and ended up with 

```
from typing import Any, Optional, TypeVar
from numpy.typing import ArrayLike,  _ShapeLike

_ScalarGeneric = TypeVar("_ScalarGeneric", bound=generic)

@overload
def beta(a: _ScalarGeneric, b: _ScalarGeneric, size: None = ...) -> float: ...
@overload
def beta(a: _ScalarGeneric, b: _ScalarGeneric, size: _ShapeLike = ...) -> ndarray: ...
@overload
def beta(a: ArrayLike, b: _ScalarGeneric, size: Optional[_ShapeLike] = ...) -> ndarray: ...
@overload
def beta(a: _ScalarGeneric, b: ArrayLike, size: Optional[_ShapeLike] = ...) -> ndarray: ...
@overload
def beta(a: ArrayLike, b: ArrayLike, size: Optional[_ShapeLike] = ...) -> ndarray: ...
```

A few questions if you have time:
1. Is there a better type for float-like but not complex? Or should we type things that will eventually be converted to a float (e.g., `a` as a float)?
2. Is there any way to specify a datatype for ArrayLike?
3. Can `_ScalarGeneric` be imported from somewhere in pyi files? 
4.. Do things like ndarray need to be imported anywhere?

File: numpy/__init__.pyi [hunk 4]
```diff
@@ -44,6 +44,25 @@ if sys.version_info >= (3, 8):
 else:
     from typing_extensions import Literal, Protocol
 
+# Ensures that the stubs are picked up
+from . import (
+    char,
+    compat,
+    core,
```
[the remark this answers is not in the document]
Comment (BvB93) [reply numpy/__init__.pyi:51]: Done as of https://github.com/numpy/numpy/pull/17104/commits/6d33a659b493b43b393e2ce58cbb9e8e40e2f583.

File: numpy/__init__.pyi [hunk 5]
```diff
@@ -44,6 +44,25 @@ if sys.version_info >= (3, 8):
...
+    ctypeslib,
```
[the remark this answers is not in the document]
Comment (BvB93) [reply numpy/__init__.pyi:52]: Done as of https://github.com/numpy/numpy/pull/17104/commits/1a3efc43575434509fb968cbe42701094d9a0dee.

I had some issues identifying the `distutils` public namespace, so for now it's still annotated as a module-level `__getattr__()`.

**Comment by BvB93** [comment]
> I would prefer if we didn't blindly annotate everything that doesn't start with an underscore (not sure if that's what you aimed for). Not typing things that are meant to be private, or things that are deprecated, would be better as a signal to users they're in the wrong place.

These annotations will definitely have to be refined in the future (they are annotated as `Any` after all).
Using the module-level `__all__` allows for some quick, and reasonably accurate, identification of all public objects. I'd say lets wait with worrying about deprecation until we start replacing the `Any`s with proper annotations.

File: numpy/ctypeslib.pyi [hunk 6]
```diff
@@ -0,0 +1,8 @@
+from typing import Any
+
+load_library: Any
+ndpointer: Any
+ctypes_load_library: Any
```
Comment (eric-wieser) [inline numpy/ctypeslib.pyi:5]: Given this has been deprecated for at least 14 years, I think we can skip it here.

https://github.com/numpy/numpy/blame/64619c55896dad1927afe78c0702a58abf0d038f/numpy/ctypeslib.py#L157

Comment (eric-wieser) [reply numpy/ctypeslib.pyi:5]: Removed entirely in #17116

File: numpy/ctypeslib.pyi [hunk 7]
```diff
@@ -1,8 +0,0 @@
-from typing import Any
-
-load_library: Any
-ndpointer: Any
-ctypes_load_library: Any
-c_intp: Any
-as_ctypes: Any
-as_array: Any
```
Comment (eric-wieser) [inline numpy/ctypeslib.pyi]: To be clear, only `ctypes_load_library` is deprecated.

Comment (BvB93) [reply numpy/ctypeslib.pyi]: Ah, my bad. Everything should be in order as of https://github.com/numpy/numpy/pull/17104/commits/2ec7afa3f4fcc5cb8792ac9f63ef38f667506c5a.

**Comment by BvB93** [comment]
@bashtage

> Is there a better type for float-like but not complex? Or should we type things that will eventually be converted to a float (e.g., a as a float)?

If https://github.com/numpy/numpy/pull/17117 gets merged then typing them as `float` will be sufficient; otherwise I'd stick to something like `_FloatLike = Union[float, np.floating, np.integer]`.

> Is there any way to specify a datatype for ArrayLike?

https://github.com/numpy/numpy/pull/16759 is currently working on making `ndarray` generic with respect to the data type, but it's very much a work in progress. For now I'd just stick to `ArrayLike`.

> Can _ScalarGeneric be imported from somewhere in pyi files?

`from numpy import _ScalarGeneric` should do the trick. Note that this won't work during runtime; only use it when type checking.

> Do things like ndarray need to be imported anywhere?

They do; you can just use `from numpy import ndarray`.

**Comment by charris** [comment]
Thanks @BvB93 .

**Pull request merged**